src/io/flutter/run/MainFile.java

```diff
@@ -0,0 +1,198 @@
+/*
+ * Copyright 2017 The Chromium Authors. All rights reserved.
+ * Use of this source code is governed by a BSD-style license that can be
+ * found in the LICENSE file.
+ */
+package io.flutter.run;
+
+import com.intellij.openapi.application.ApplicationManager;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.roots.ProjectRootManager;
+import com.intellij.openapi.util.io.FileUtil;
+import com.intellij.openapi.util.text.StringUtil;
+import com.intellij.openapi.vfs.LocalFileSystem;
+import com.intellij.openapi.vfs.VirtualFile;
+import com.intellij.psi.PsiFile;
+import com.intellij.psi.PsiManager;
+import com.intellij.psi.util.PsiTreeUtil;
+import com.jetbrains.lang.dart.DartFileType;
+import com.jetbrains.lang.dart.psi.DartFile;
+import com.jetbrains.lang.dart.psi.DartImportStatement;
+import com.jetbrains.lang.dart.util.DartResolveUtil;
+import io.flutter.FlutterBundle;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Arrays;
+import java.util.stream.Stream;
+
+/**
+ * The location of a Dart file containing a main() method for launching a Flutter app.
+ * <p>
+ * It has been checked for errors, but the check might not have succeeded.
+ */
+public class MainFile {
+
+  @NotNull
+  private final VirtualFile file;
+
+  @NotNull
+  private final VirtualFile appDir;
+
+  private final boolean flutterImports;
+
+  private MainFile(@NotNull VirtualFile file, @NotNull VirtualFile appDir, boolean flutterImports) {
+    this.file = file;
+    this.appDir = appDir;
+    this.flutterImports = flutterImports;
+  }
+
+  /**
+   * Returns the Dart file containing main.
+   */
+  @NotNull
+  public VirtualFile getFile() {
+    return file;
+  }
+
+  /**
+   * Returns the closest ancestor directory containing a pubspec.yaml, BUILD, or .packages file.
+   */
+  @NotNull
+  public VirtualFile getAppDir() {
+    return appDir;
+  }
+
+  /**
+   * Returns true if the file has any direct flutter imports.
+   */
+  public boolean hasFlutterImports() {
```

Comment:
It is unused now, but I'd like to change FlutterRunConfigurationProducer to use MainFile in a future CL.